oauth2/oauth2-resource-server/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/resourceserver/OAuth2ResourceServerConfigurer.java

```diff
@@ -0,0 +1,159 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.config.annotation.web.configurers.oauth2.resourceserver;
+
+import org.springframework.security.config.annotation.web.HttpSecurityBuilder;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.oauth2.core.AbstractOAuth2Token;
+import org.springframework.security.oauth2.core.OAuth2TokenVerifier;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.resourceserver.web.BearerTokenResolver;
+
+/**
+ * Example configuration:
+ *
+ * oauth2().resourceServer().accessToken()
+ *     .formats()
+ *         .jwt().and()
+ *     .verifiers()
+ *         .signature().keys("http://jwk.url")
+ *
+ * Or:
+ *
+ * oauth2().resourceServer().accessToken()
+ *     .formats()
+ *         .jwt()
+ *             .processor(auth0AccessTokenProcessor())
+ *             .and()
+ *         .opaque()
+ *             .processor(auth0AccessTokenProcessor())
+ *             .and()
+ *         .and()
+ *     .verifiers()
+ *          .addVerifier(claims -> {
+ *              if ( claims.get("iss") == null ) {
+ *                  throw new OAuth2AuthenticationException(...);
+ *              }
+ *          })
+ *
+ * @author Josh Cummings
+ */
+public class OAuth2ResourceServerConfigurer<B extends HttpSecurityBuilder<B>> extends
+		AbstractHttpConfigurer<OAuth2ResourceServerConfigurer<B>, B> {
+
+	private AccessTokenFormatsConfigurer accessTokenFormatsConfigurer;
+	private AccessTokenVerifiersConfigurer accessTokenVerifiersConfigurer;
+
+	public OAuth2ResourceServerConfigurer<B> bearerTokenResolver(BearerTokenResolver resolver) {
+		return this;
+	}
+
+	public AccessTokenConfigurer accessToken(OAuth2TokenVerifier... verifiers) {
+		return new AccessTokenConfigurer();
+	}
+
+	public class AccessTokenConfigurer {
+		public AccessTokenVerifiersConfigurer verifiers() {
+			return new AccessTokenVerifiersConfigurer();
+		}
+
+		public AccessTokenFormatsConfigurer formats() {
+			return new AccessTokenFormatsConfigurer();
+		}
+
+		public OAuth2ResourceServerConfigurer<B> and() {
+			return OAuth2ResourceServerConfigurer.this;
+		}
+	}
+
+	public class AccessTokenVerifiersConfigurer {
+		public SignatureVerificationConfigurer signature() {
+			return new SignatureVerificationConfigurer();
+		}
+
+		public EncryptionVerificationConfigurer encryption() {
+			return new EncryptionVerificationConfigurer();
+		}
+
+		public AccessTokenVerifiersConfigurer addVerifier(OAuth2TokenVerifier verifier) {
+			return this;
+		}
+
+		public AccessTokenConfigurer and() {
+			return null;
+		}
+	}
+
+	public class SignatureVerificationConfigurer {
+		public SignatureVerificationConfigurer keys(String uri) {
```

Comment:
What if the user wanted an option to configure a PEM encoded PK which is also a `String`. This naming would collide.

Reply:
Are you saying that folks may want to indicate the raw key as a string? I'd more expect to see a `byte[]` if I wanted to hand in a raw value.

We can certainly support that, though I was imagining handing in an instance of `Key`, `SecretKey`, or the like.

I'm hoping to overload the `keys` method instead of doing something like `keysUri`, `keysValue`, etc.. I think that having all of them have the same name communicates more clearly that you get to pick one (though I'm happy to be swayed).

So, what about:

```java
keys(URI location)
keys(byte[] encoded)
keys(Key key)
keys(Map<String, Key> keysByVersion)
```

Auth0 also has an interface that I like which we could pattern after:

```java
keys(KeyProvider provider)
```

Reply:
Of course, I think this refers to a ticket that we haven't actually implemented yet, and perhaps our thoughts will change once that is complete.

Reply:
> Are you saying that folks may want to indicate the raw key as a string? I'd more expect to see a byte[] if I wanted to hand in a raw value.

Yes, Spring Boot 1.5.x supports this - [see](https://docs.spring.io/spring-boot/docs/1.5.12.RELEASE/reference/htmlsingle/#boot-features-security-custom-user-info)

```
security:
    oauth2:
        resource:
            jwt:
                keyValue: |
                    -----BEGIN PUBLIC KEY-----
                    MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKC...
                    -----END PUBLIC KEY-----
```

Reply:
> We can certainly support that, though I was imagining handing in an instance of Key, SecretKey, or the like.

Agreed, I see us initially supporting `java.security.PublicKey` and `String` for the verification keys.

Reply:
> I'm hoping to overload the keys method

I would prefer the overloading as well but how does the user know they are configuring a verification key vs. a signer key? This needs to be clear to the user. Obviously `PublicKey` and `PrivateKey` would make it clear but than it would still be fuzzy when the user wanted to configure key encryption (for example, JWE).

`keys().key(java.security.PrivateKey)`

It's not clear if this is for configuring the Signing key or Decryption key?

We might need to be more explicit than general overloading, for example:

```
keys()
    .verifyKey(rsaKey1)		// type java.security.PublicKey - used to verify signature
	.decryptKey(rsaKey2)	// type java.security.PrivateKey - used to decrypt token content
```

Something to consider for the design...

Reply:
Fair. Good feedback.

Reply:
> how does the user know they are configuring a verification key vs. a signer key?

From context. If we are configuring a resource server, we need decryption and verification:

```java
verifiers()
    .signature().keys()
    .encryption().keys()
````

makes it clear that the keys after `signature` are for the signature (verification), for example.

Reply:
Yes it does make it clear but I'm not sure it makes sense to have the `keys()` defined under the `verifiers()`. An `OAuth2TokenVerifier` is responsible for validating the claims/attributes of the access token after the signature has been verified or content has been decrypted. Now I'm wondering if the name `OAuth2TokenVerifier` is misleading?